File: 101_symmetric_tree/isSymmetric.py

# recursive
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root: return True
        def ifsymmetric(root1, root2):
            if not root1 or not root2:
                return root1 == root2

            if root1.val != root2.val:
                return False
            else:
                return ifsymmetric(root1.left, root2.right) and ifsymmetric(root1.right, root2.left)

        res = ifsymmetric(root.left, root.right)
        return res


# non-recursive

File: 101_symmetric_tree/isSymmetric.cpp
#include <iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
	public:
    	bool isSymmetric(TreeNode* root) {
            return ifSymmetric(root, root);
    	}
    private:
        bool ifSymmetric(TreeNode* a, TreeNode* b) {
            if(!a || !b) return a == b;
            return a -> val == b -> val && \
                   ifSymmetric(a -> left, b -> right) && \
                   ifSymmetric(a -> right, b -> left);
        }
};
